### PR TITLE
test: add comprehensive edge case tests for the range function in Drasi Core

### PR DESCRIPTION
**Pull request by IronJam11**
# Description

This PR adds comprehensive unit test coverage for the `range` list function, focusing on edge cases, boundary conditions, and invalid argument handling.

The tests document and validate existing behavior for negative ranges, step handling, invalid inputs, and error semantics, improving confidence in correctness and preventing future regressions.

## New Tests Added (11 total)

- **test_list_range_invalid_arguments** – Tests `range(2, '18')` rejects string as second argument with `InvalidArgument(1)`
- **test_list_range_negative_numbers** – Tests range crossing zero from `-5` to `5`, returns 11 elements
- **test_list_range_all_negative** – Tests entirely negative range from `-10` to `-5`, returns 6 negative integers
- **test_list_range_start_greater_than_end** – Tests `range(10, 5)` returns empty list when start > end
- **test_list_range_single_element** – Tests `range(5, 5)` returns single-element list `[5]`
- **test_list_range_negative_with_step** – Tests `range(-10, 10, 5)` with negative start and step value
- **test_list_range_large_step** – Tests `range(0, 10, 100)` where step exceeds range, returns only the start element
- **test_list_range_invalid_step_argument** – Tests `range(0, 10, '2')` rejects string as step with `InvalidArgument(2)`
- **test_list_range_float_arguments** – Tests `range(0.5, 10.5)` rejects float arguments with `InvalidArgument(0)`
- **test_list_range_zero_step** – Tests `range(0, 10, 0)` panics with `assertion failed: step != 0` (`should_panic`)
- **test_list_range_negative_step** – Tests `range(0, 10, -1)` casts step to large `usize`, returns `[0]`

**Total range tests:** 14 (3 existing + 11 new)

## Type of change

- This pull request is a minor refactor, code cleanup, test improvement, or other maintenance task and doesn't change the functionality of Drasi.

## Related Issue

Related to: #245

### CODE REVIEW

File: core/src/evaluation/expressions/tests/list_functions.rs [hunk 1]
```diff
@@ -288,6 +288,280 @@ async fn test_list_range_invalid_arg_count() {
...
+    let function_registry = create_list_expression_test_function_registry();
+    let ari = Arc::new(InMemoryResultIndex::new());
+    let evaluator = ExpressionEvaluator::new(function_registry.clone(), ari.clone());
+
+    let variables = QueryVariables::new();
+    let context =
+        ExpressionEvaluationContext::new(&variables, Arc::new(InstantQueryClock::new(0, 0)));
+    // This will panic with "assertion failed: step != 0" error
+    let _ = evaluator.evaluate_expression(&context, &expr).await;
+}
```
[the remark this answers is not in the document]
Comment (IronJam11) [reply core/src/evaluation/expressions/tests/list_functions.rs:539]: What are your views on this @amansinghoriginal @danielgerlag @agentofreality ?? Should i add a validation check in the function.

File: core/src/evaluation/expressions/tests/list_functions.rs [hunk 2]
```diff
@@ -288,6 +288,280 @@ async fn test_list_range_invalid_arg_count() {
...
+    // Negative step is cast to usize, resulting in a very large step
+    // This should result in only the first element
+    let result = evaluator
+        .evaluate_expression(&context, &expr)
+        .await
+        .unwrap();
+    assert_eq!(
+        result,
+        VariableValue::List(vec![VariableValue::Integer(Integer::from(0))])
+    );
```
[the remark this answers is not in the document]
Comment (IronJam11) [reply core/src/evaluation/expressions/tests/list_functions.rs:562]: and this as well?

File: core/src/evaluation/expressions/tests/list_functions.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (danielgerlag) [reply core/src/evaluation/expressions/tests/list_functions.rs:539]: yes, it should return an error that can bubble up, without a panic.

File: core/src/evaluation/expressions/tests/list_functions.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (danielgerlag) [reply core/src/evaluation/expressions/tests/list_functions.rs:562]: this as well

File: core/src/evaluation/expressions/tests/list_functions.rs [hunk 3]
```diff
@@ -288,6 +288,280 @@ async fn test_list_range_invalid_arg_count() {
...
+        result,
+        EvaluationError::FunctionError(FunctionError {
+            function_name: _,
+            error: FunctionEvaluationError::InvalidArgument(0)
+        })
+    ));
+}
+
+#[tokio::test]
+#[should_panic(expected = "assertion failed: step != 0")]
```
Comment (amansinghoriginal) [inline core/src/evaluation/expressions/tests/list_functions.rs:523]: `#[should_panic]` is the right tool when a panic is *intentional* - but here, should it be? If a user writes a query with `range(0, 10, 0)`, should that crash the entire query engine process? Take a look at how the function handles other invalid inputs (like string arguments) and apply the same philosophy here. The fix belongs in `range.rs` - this test should then assert the corrected behavior.

File: core/src/evaluation/expressions/tests/list_functions.rs [hunk 4]
```diff
@@ -288,6 +288,280 @@ async fn test_list_range_invalid_arg_count() {
...
+    let expr = drasi_query_cypher::parse_expression(expr).unwrap();
+
+    let function_registry = create_list_expression_test_function_registry();
+    let ari = Arc::new(InMemoryResultIndex::new());
+    let evaluator = ExpressionEvaluator::new(function_registry.clone(), ari.clone());
+
+    let variables = QueryVariables::new();
+    let context =
+        ExpressionEvaluationContext::new(&variables, Arc::new(InstantQueryClock::new(0, 0)));
+    // Negative step is cast to usize, resulting in a very large step
```
Comment (amansinghoriginal) [inline core/src/evaluation/expressions/tests/list_functions.rs:553]: Your comment here shows you understand exactly what's happening - `-1` wraps to a huge `usize`. The question to ask is: would any user calling `range(0, 10, -1)` expect `[0]` back? Think about what the function *should* do with a negative step, and make it do that. The same fix in `range.rs` that addresses the zero-step case can handle this one too.

**Review by amansinghoriginal — CHANGES_REQUESTED** [review]
Really nice work, @IronJam11! The coverage you've added for negative ranges, boundary conditions, type validation, and large steps is exactly what we needed - 9 of these 11 tests are solid and ready to go.

Your comments in the last two tests show you already found two bugs in the implementation - the zero-step panic and the negative-step casting issue. That's great investigative work, and it's exactly what the [Stretch Goal](https://github.com/drasi-project/drasi-core/issues/245) in the issue was hoping you'd discover.

The next step is to go from documenting these bugs to **fixing them**. As @danielgerlag mentioned, both cases should return an error rather than panicking or producing surprising output. Take a look at how the implementation in `range.rs` handles the `step` value before it reaches `step_by()`, and think about what validation is missing. Once you fix the implementation, update these two tests to assert the correct behavior - you already have great examples of error-asserting tests in this same PR to use as a pattern.

**Review by amansinghoriginal — CHANGES_REQUESTED** [review]
Requesting Changes

**Comment by IronJam11** [comment]
Sure @amansinghoriginal  @danielgerlag I will work on fixing them as well. Sorry for the late reply - I was sick the past few days.

File: core/src/evaluation/expressions/tests/list_functions.rs (hunk 4 above)
Comment (IronJam11) [reply core/src/evaluation/expressions/tests/list_functions.rs:553]: @amansinghoriginal  I'm working on the fixes now. I'll implement a validation error for step == 0 to prevent the panic.

Regarding negative steps: 

Use Cases for Negative Steps:
Descending sequences: range(10, 0, -1) → [10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0]
- Countdown timers in queries
- Reverse iteration through indices
- Priority ordering (high to low)
- Compatibility with Neo4j Cypher: The range() function in Neo4j supports negative steps for descending ranges, which Drasi should match for compatibility.

Query flexibility: Users might want to generate descending number sequences for:

- Time-based windowing (most recent to oldest)
- Ranking systems (top to bottom)
- Reverse chronological data processing

Comment (IronJam11) [reply core/src/evaluation/expressions/tests/list_functions.rs:553]: @amansinghoriginal pls have a look at this. thank you
